### PR TITLE
Put Dockerfile in its own docker folder

### PR DESCRIPTION
**Pull request by mathbunnyru**
It would be better to put the Dockerfile in the separate folder.

1. Users might (and they probably will for some config files) add some additional files during building their image - this will allow them to put everything in one separate folder, which keeps the project clean.
2. It allows to have a better workflow, which is only triggered when relevant files are changed.
3. This also allows us to remove .dockerignore. It reduces the complexity of the project for the end user (and for us as well).

### CODE REVIEW

**Comment by consideRatio** [comment]
@mathbunnyru the fixes are not motivated, and while I can find some motivations myself its always helpful to know ahead of time what your motivation was for the changes.

My perception is that you think it reduces complexity if you have a dedicated folder for the Dockerfile's build contex, which helps reduce the .dockerignore file for example and trigger github workflows more easily on changes to that build context. That is at least what I think and I'm positive to the change.

I'm considering calling the folder something else than "docker". I think it isn't very descriptive. Sure it has a dockerfile in it but hmmm.. Consider for example jupyter/docker-stacks, we have folders named like the image we will publish. I think `image` could be more helpful to convey what it is about. It could be a variable name as well, but I think it can increase the complexity too much to motivate it. I'm thinking its both easier to overview a cookiecutter template repo and maintain it if not everything is a variables.

### Summary

- The change idea makes sense to me!
- The folder name - perhaps `image` instead? WDYT?
- An explicit change motivation in PR is helpful not just for review, but historical inspection for anyone - even yourself in the future.

**Comment by mathbunnyru** [comment]
> The change idea makes sense to me!

Thank you, @consideRatio! 👍 

> The folder name - perhaps image instead? WDYT?

You're right, this looks better.
I've seen other projects containing docker folder, that's why I called it docker, but usually, it's not their main goal.
Here it is.

> An explicit change motivation in PR is helpful not just for review, but historical inspection for anyone - even yourself in the future.

Sure, I will add the description.

File: {{cookiecutter.stack_name}}/.github/workflows/docker.yml [hunk 1]
```diff
@@ -4,14 +4,27 @@ name: Build, test, and publish Docker Images
...
+      - "Makefile"
+      - "requirements-dev.txt"
   push:
     branches:
       - master
       - main
-    paths-ignore:
-      - "*.md"
+    paths:
+      - ".github/workflows/docker.yml"
```
Comment (consideRatio) [inline {{cookiecutter.stack_name}}/.github/workflows/docker.yml:20]: This is feature creep for the PR, but this is an example of a name that I think isn't as helpful as `"build-image.yaml"` or `"test-build-push-image.yaml"` or similar.

**Review by consideRatio — APPROVED** [review]

File: {{cookiecutter.stack_name}}/.github/workflows/docker.yml (hunk 1 above)
Comment (mathbunnyru) [reply {{cookiecutter.stack_name}}/.github/workflows/docker.yml:20]: I think it follows the name of the "upstream" project.
https://github.com/jupyter/docker-stacks/tree/master/.github/workflows

But I think your naming makes more sense and I will probably do it soon in separate PR(s).

**Pull request merged**